### PR TITLE
Compression updates v2

### PR DESCRIPTION
**Pull request by alphmth**
**Please check if the PR fulfills these requirements**
- [x] Followed the [Contributing](https://github.com/jaredwray/keyv/blob/main/CONTRIBUTING.md) guidelines.
- [x] Tests for the changes have been added (for bug fixes/features) with 100% code coverage.
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
compression updates

### CODE REVIEW

File: packages/compress-brotli/src/index.js [hunk 1]
```diff
@@ -3,28 +3,25 @@ const compressBrotli = require('compress-brotli');
...
-		};
-
-		const {compress, decompress, serialize, deserialize} = compressBrotli(this.opts);
+		this.opts = options;
+		this.brotli = compressBrotli(this.opts);
+	}
 
-		this.opts.compress = compress;
-		this.opts.serialize = async ({value, expires}) => serialize({value: await compress(value), expires});
+	compress(value) {
```
Comment (jaredwray) [inline packages/compress-brotli/src/index.js]: can we make this async?

File: packages/compress-brotli/src/index.js [hunk 2]
```diff
@@ -3,28 +3,25 @@ const compressBrotli = require('compress-brotli');
...
+		return this.brotli.compress(value, this.opts);
+	}
 
-		this.opts.decompress = decompress;
-		this.opts.deserialize = async data => {
-			const {value, expires} = deserialize(data);
-			return {value: await decompress(value), expires};
-		};
+	decompress(data) {
```
Comment (jaredwray) [inline packages/compress-brotli/src/index.js]: can we make this async?

File: packages/compress-gzip/src/index.d.ts [hunk 3]
```diff
@@ -1,19 +1,11 @@
...
-declare namespace KeyvGzip {
-	interface Options {
-		compress: (...args: any[]) => void;
-		decompress: (...args: any[]) => void;
-	}
+	constructor(options?);
+	compress(value: any);
+	decompress(value: any);
+	serialize(value: any);
+	deserialize(value: any);
```
Comment (jaredwray) [inline packages/compress-gzip/src/index.d.ts]: please make these async?

File: packages/compress-gzip/src/index.js [hunk 4]
```diff
@@ -7,23 +7,23 @@ class KeyvGzip {
...
 
-		this.opts.compress = pako.deflate;
-		this.opts.decompress = pako.inflate;
-		this.opts.serialize = async ({value, expires}) => JSONB.stringify({value: await this.opts.compress(value, this.opts), expires});
-		this.opts.deserialize = async data => {
-			const {value, expires} = JSONB.parse(data);
-			const value_ = await this.opts.decompress(value, this.opts);
-			return {value: value_, expires};
-		};
+	compress(value) {
```
Comment (jaredwray) [inline packages/compress-gzip/src/index.js]: async please

File: packages/compress-gzip/src/index.js [hunk 5]
```diff
@@ -7,23 +7,23 @@ class KeyvGzip {
...
+		return pako.deflate(value, this.opts);
+	}
+
+	decompress(value) {
```
Comment (jaredwray) [inline packages/compress-gzip/src/index.js]: async please

File: packages/keyv/src/index.d.ts [hunk 6]
```diff
@@ -85,7 +85,12 @@ declare namespace Keyv {
 		/** Specify an adapter to use. e.g `'redis'` or `'mongodb'`. */
 		adapter?: 'redis' | 'mongodb' | 'mongo' | 'sqlite' | 'postgresql' | 'postgres' | 'mysql' | undefined;
 		/** Enable compression option **/
-		compress?: Record<string, unknown> | undefined;
+		compression?: CompressionAdapter | undefined;
+	}
+
+	interface CompressionAdapter {
+		compress: (value: any) => any;
+		decompress: (value: any) => any;
```
Comment (jaredwray) [inline packages/keyv/src/index.d.ts]: can we add in `options` parameter?

**Review by jaredwray — CHANGES_REQUESTED** [review]

File: packages/keyv/src/index.d.ts (hunk 6 above)
Comment (stropicall) [reply packages/keyv/src/index.d.ts]: @jaredwray Earlier we are supporting `options` parameter but you said pass the `options` parameter in constructor

Comment (jaredwray) [reply packages/keyv/src/index.d.ts]: I would like it to be passed via a property or as a parameter in the constructor. Either way it should work and we should test for that.

Comment (stropicall) [reply packages/keyv/src/index.d.ts]: ok @jaredwray I will do that

**Comment by alphmth** [comment]
> @alphmth - how would we separate this out so that serialize, deserialized could exist without being part of the compression package or maybe an option that if those functions exist we use it otherwise we use what is set?

In the `set` or `get` method we are using `serialize` and `deserialize` methods. To make compression work we need to add compress/decompress method in the serialize/deserialize method. If we make them optional how we will use compression in main `keyv` project?

**Comment by jaredwray** [comment]
@alphmth - this is starting to come together. One other thought is to create another export of the test-suite library that does validation of the compression packages so third parties can do that. Can you add that in?

**Comment by alphmth** [comment]
@jaredwray do we need to add `serialize/deserialize` as optional? In that case, the user will have to use `compress/decompress` methods like this

```
const gzip = new KeyvGzip(); 
const compressedValue = gzip.compressed(value);
const keyv = new Keyv({store: new Map()});
await keyv.set('test-key', compressedValue, 1000);

//get compressed value
const decompressedValue = await gzip.decompressed(await keyv.get('test-key));

```

**Comment by jaredwray** [comment]
> @jaredwray do we need to add `serialize/deserialize` as optional? In that case, the user will have to use `compress/decompress` methods like this
> 
> ```
> const gzip = new KeyvGzip(); 
> const compressedValue = gzip.compressed(value);
> const keyv = new Keyv({store: new Map()});
> await keyv.set('test-key', compressedValue, 1000);
> 
> //get compressed value
> const decompressedValue = await gzip.decompressed(await keyv.get('test-key));
> ```

Lets go with the current architecture for right now and then will take feedback.

**Pull request merged**